### ISSUE TITLE
Move to esbuild

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/ex_doc#1542]
We will probably need npm still, for handlebars, but it should streamline things considerably.

**Comment by sigu** [comment]
I can get started in this, what is the procedure?

**Comment by josevalim** [comment]
The next step would be to submit a PR that replaces webpack by esbuild. Please let me know if you have any questions. :)

**Comment by osbre** [comment]
Is this a complete webpack removal or adding [esbuild-loader](https://github.com/privatenumber/esbuild-loader) to speed it up?

**Comment by DavidOliver** [comment]
@sigu, it sounds like you're taking this, but let me know if not as I'd be happy to have a go in between other bits over the next few days.

@josevalim, I wondered if this would be a good time to remove Less for CSS. I tend to use plain CSS now as browser support for custom properties (variables) is good.

**Comment by josevalim** [comment]
Ah, I forgot about less. I think we can definitely consider removing it. Perhaps before moving to esbuild. Do you want to analyze how doable doing so is? Thanks!

**Comment by DavidOliver** [comment]
Sure - will do at some point this week hopefully, unless @sigu chooses to look at that part before I get to it. :)

**Comment by sigu** [comment]
Hello @DavidOliver I will have a look at how doable it is in the next few hours and share my findings. You can also have a look in parallel

**Comment by cw789** [comment]
Removing less is in my opinion a good idea.

**Comment by DavidOliver** [comment]
> Hello @DavidOliver I will have a look at how doable it is in the next few hours and share my findings. You can also have a look in parallel

Hi @sigu. What's the state of play with this your end? I started to have a quick look through the less files and saw that some reworking would be required with regards to nested selectors and content variables. I'd be happy to take a look at handling the switch to plain CSS next week if you'd like?

**Comment by DavidOliver** [comment]
I've been away but have now made a start on removing Less.

**Comment by DavidOliver** [comment]
@josevalim, while I'm going through all the CSS, should I remove styles and assets that were added to support now-out-of-scope(?) browsers such as IE and old versions of current browsers? E.g., removing all font formats except for woff2.

**Comment by josevalim** [comment]
Sounds good to me!

**Comment by moogle19** [comment]
@DavidOliver 
I played around with the `lessc` command line tool and got a working version with less completly removed (see https://github.com/moogle19/ex_doc/tree/less_less). It didn't produce the most beautiful css but could be a good starting point.

**Comment by DavidOliver** [comment]
@moogle19, cool - thanks. Based on a very quick look, my approach is a more manual (and slow!) one, giving attention to each included file, and I'm preserving comments and the specifying of alpha values in the main CSS by separating out h, s, and l values in the custom properties where necessary. I'll carry on in this vein with long-term maintenance in mind, but it's good to see you've got it all working.

I hope to carry on and update here this week.

**Comment by DavidOliver** [comment]
Sorry for the long delay. I've since moved, and I'm hoping to be in a position to resume work on projects at some point next week.

**Comment by DavidOliver** [comment]
I've now picked this up again, and have rebased and incorporated the improvements to dark mode made since I started removing Less. I'll continue working my way through the remaining Less files.

(Sorry again!)

**Comment by DavidOliver** [comment]
Initial run-through mostly complete. I'm now reworking how custom properties are used and tweaking some selector blocks in order to cut down on highly repetitive selector rules and to avoid the need to redefine colours in the 'night' files.

E.g., we can define most colour custom properties for light and dark upfront as follows:

```css
:root {
  --blockquoteBackground: hsl(0, 0%, 97%);
  --warningBackground: hsl(33, 100%, 97%);
}

body.dark {
  --blockquoteBackground: hsl(240, 5%, 18%);
  --warningBackground: hsl(40, 67%, 79%);
}
```

Then we only need to set most color properties once:

```css
.content-inner blockquote.warning {
  background-color: var(--warningBackground);
}
```

And no longer need to additionally set the dark theme color properties as at present:

```css
body.dark .content-inner blockquote.warning {
  background-color: #edd5a5;
}
```

This and other adjustments are taking some time to work through, especially as I'm not already familiar with the ExDoc epub and HTML output, but I think it will be worth it as there will be less CSS and it should be easier to maintain/update.

**Comment by DavidOliver** [comment]
PR to remove Less CSS submitted. I can also have a go at moving to esbuild, but I thought I should get the removal of Less CSS in before any more styling changes are made if possible. :)

**Comment by DavidOliver** [comment]
Status update: the main/basic build using esbuild is working. :partying_face:

Handlebars was the biggest issue; I worked around not being able to get either of the available loaders to work by setting up precompilation of the templates, which I think in theory saves some processing power in the browser.

I'm yet to check that things like the optional logo and extra assets are working, but will do before submitting the PR.

For this move to esbuild, should I include the changes to files in `formatters` in my commit? I ask as the readme says not to when contributing.

**Comment by josevalim** [comment]
Please don't, I will do it right after :) Thank you! :heart:
